### DISCUSSION TITLE
Is it possible to draw molecule with bond lengths and angles on top?

### DISCUSSION THREAD

**Discussion by cdelv** [rdkit/rdkit#4832] [Q&A]
Hi, is it possible to draw a molecule out of an XYZ coordinate file and make the program automatically put the bond lengths or the bond angles on the output picture? I added an example picture. 
![Screenshot from 2021-12-27 14-38-35](https://user-images.githubusercontent.com/54986485/147516691-85013a65-19cf-4a4d-9874-5d0d333994bc.png)

**Comment by greglandrum** [comment]
@cdelv : this is a fun one. :-)
Bond labels are pretty easy: it's possible to add notes to each bond in the molecule which are drawn when the molecule is drawn.
Here's a simple example which labels bond lengths:

```
m = Chem.MolFromSmiles('c1nc[nH]c1F')
mh = Chem.AddHs(m)
# create a 3D conformer for a copy:
mh_cp = Chem.Mol(mh)
rdDistGeom.EmbedMolecule(mh_cp)

conf = mh_cp.GetConformer()

for bnd in mh_cp.GetBonds():
    bl = rdMolTransforms.GetBondLength(conf,bnd.GetBeginAtomIdx(),bnd.GetEndAtomIdx())
    # set the bond note on the other molecule, which we'll draw:
    mh.GetBondWithIdx(bnd.GetIdx()).SetProp('bondNote',f'{bl:.2f}')
```
This produces the following output when you draw it with a `MolDraw2D` object (or in the notebook):
![image](https://user-images.githubusercontent.com/540511/147526152-6d5cfd21-e755-4400-8e5c-5ac8c36b1988.png)

You can change the color palette to be black and white, like your example image, and draw the atoms as circles, but you can't get exactly the same drawing style you show since it's not possible to specify separate drawing colors for the outline and interior of atoms. Still, it's not terrible:
```
rads = {}
colors = {}
for atom in mh.GetAtoms():
    if atom.GetAtomicNum()!=1:
        rads[atom.GetIdx()]=0.2
    else:
        rads[atom.GetIdx()]=0.1
    if atom.GetAtomicNum() !=6:
        colors[atom.GetIdx()] = (0.8,0.8,0.8)
    else:
        colors[atom.GetIdx()] = (0.,0.,0.)

d2d = Draw.MolDraw2DCairo(250,200)
d2d.drawOptions().continuousHighlight=False
d2d.drawOptions().noAtomLabels=True
d2d.drawOptions().splitBonds=False
d2d.drawOptions().fillHighlights=True
d2d.drawOptions().useBWAtomPalette()
d2d.DrawMolecule(mh,highlightAtoms=range(mh.GetNumAtoms()),highlightBonds=[],
                 highlightAtomColors=colors,highlightAtomRadii=rads)
d2d.FinishDrawing()
```
Produces this:
![image](https://user-images.githubusercontent.com/540511/147527233-dda1fea2-677b-40a5-945b-783857e92b76.png)

Showing bond angles is harder: each atom also can have an `atomNote` associated with it using the same `SetProp()` mechanism, but you would also need some way to specify which other atoms are associated with the note. And, at the moment, there's only one `atomNote` per atom.

**Reply by DavidACosgrove** [reply]
Another problem with using the atomNote to show the bond angle is that you, the user, have no control over where it is drawn - the code just tries to find somewhere near the atom that doesn't clash with anything else in the picture.  So it is more than likely that if you attempt to add an atomNote showing a bond angle it will end up "inside" the wrong angle.  It may be possible to do something with the lower-level MolDraw2D drawing functions.  I will give it some thought.  There is code to draw a string at a particular place, and also to draw an arc which in principle could be pressed into action.  In practice it will likely be fiddly to get them in the right place.

**Reply by cdelv** [reply]
@greglandrum Something appears to be wrong with the code. I used the code you posted and got almost all the bond lengths wrong.
![Screenshot from 2021-12-28 12-56-17](https://user-images.githubusercontent.com/54986485/147593776-4e97a4ef-632f-4555-b50f-6ff78cddb695.png)

 I used Jmol and Gdis to corroborate the bonds and in both cases they are wrong. I'm plotting with Jmol the same .mol file I used to load the molecule into RDKit.
` m = Chem.MolFromMolFile('A-T2.mol', removeHs=False)`

Why am I getting the wrong bond lengths? Is there a bug in the code or do I have to tell RDKit to load the molecule as 3D before calculating the lengths?

**Reply by DavidACosgrove** [reply]
I think the issue here could well be that, without a 3D structure to work from, @greglandrum used EmbedMolecule to generate one.  Did you remove that line from your script?

**Reply by cdelv** [reply]
@DavidACosgrove, no I didn't. I ended up extracting de coordinates of the .mol file and using numpy.linalg.norm to get the distances. I used the atom Ids to calculate the lengths.

**Comment by DavidACosgrove** [comment]
@cdelv This has lightened up a miserable wet Bank Holiday.  You could try something like inserting the following in Greg's script:

```
d2d.DrawMolecule(mh,highlightAtoms=range(mh.GetNumAtoms()),highlightBonds=[],
                 highlightAtomColors=colors,highlightAtomRadii=rads)
mh_conf = mh.GetConformer()
for atom in mh_cp.GetAtoms():
    nbrs = [nbr for nbr in atom.GetNeighbors()]
    if len(nbrs) > 1:
        ang = rdMolTransforms.GetAngleDeg(conf, nbrs[0].GetIdx(), atom.GetIdx(), nbrs[1].GetIdx())
        pos = mh_conf.GetAtomPosition(atom.GetIdx())
        d2d.DrawString(f'{ang:6.2f}', Geometry.Point2D(pos.x, pos.y))
d2d.FinishDrawing()

```

Which produces the rather ugly looking:

![image](https://user-images.githubusercontent.com/9198870/147571348-a7e224e5-dba3-4522-a4ec-eb41b8a28a43.png)


This could be improved by drawing the string at the mean of the 3 coords defining the angle, for example, rather than putting it on top of the middle atom of the angle.  There is a C++ function to change the scale of the font (setFontScale) but it doesn't appear to be exposed in the Python layer.  The code that draws the annotations calls it internally which is why those numbers are half the size.   Also, there is a 3rd parameter to MolDraw2D.DrawString() that defines alignment of the text - it defaults to MIDDLE but would probably be best as START for these purposes.  I couldn't work out how to turn the C++ RDKit::TextAlignType::START into whatever Python understands.  No doubt @greglandrum could advise.

The function MolDraw2D.DrawArc() takes a centre, radius and start and finish angles, and that could be used to draw an arc to show what the angle number refers to.  I leave that as an exercise for the reader :-).

I fear this is not a great solution to your problem.  I am currently re-factoring the drawing code, and as part of that I can at least expose setFontScale which will help.  That won't be until the next release, though.   If you need something sooner, then I fear your best bet might be to cross over to the dark size and write some C++.

**Reply by cdelv** [reply]
Thanks, @DavidACosgrove, and @greglandrum. Your answers are great. I have been searching the internet for a piece of software capable of doing this for months. I won't accept the answer yet, I'll wait for the next update and also will dig in on the C++ if I have some time. 

I have a little question, both of you, and also in the RDKit documentation, there is always one molecule per drawing. Is it possible to draw different molecules on the same drawing? Like for example the adenine thymine pair I posted? Don't bother with a complicated example, two water molecules will be enough. 

Thanks.

**Reply by greglandrum** [reply]
Here's some progress on drawing angles:
![image](https://user-images.githubusercontent.com/540511/147633415-f7f8244b-4ec3-469c-a4fa-085f4dc45935.png)
and here's a gist with the code for that:
https://gist.github.com/greglandrum/bf7a0d6b91ee2afb32724fc22ad8faf0
**NOTE:** The code in the gist will not currently work for you. It requires a couple of bugs to be fixed in the RDKit core first (it turns out that the code to draw primitives to the molecule canvas is not particularly well tested...). These will hopefully be in the next RDKit patch release.

**Reply by DavidACosgrove** [reply]
This [blog post](https://asteeves.github.io/blog/2015/01/14/editing-in-rdkit/) shows how to combine 2 molecules.  You can also use a dot-connected SMILES string ('c1ccccc1O.ClCCCC') to create one directly.

**Reply by cdelv** [reply]
@DavidACosgrove, Thanks. Because I'm working with a .mol file instead of the smiles, what I did was to add the hydrogen bond manually with `new_mol.AddBond(8,25,Chem.rdchem.BondType('21'))`. Previously I created a `Chem.RWMol` out of the `Chem.MolFromMolFile` and after I was done adding the bonds, I converted it back to a `Chem.Mol`. Type 21 stands for bond zero, so it gives me a nice dotted line. Because I added these new bonds, I had to skip the `GetConformer` part because it was complaining about bond orders. Although, this wasn't a problem because I had already replaced the function of the bond lengths with `numpy.linalg.norm`. I got some nice results, here is one of them. 
![test](https://user-images.githubusercontent.com/54986485/147681129-5a291d34-4523-4415-a443-9c64c85b2c4a.png)

Again, thanks for all your help. I'm really looking forward to the angle code.
